<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link href="//fonts.googleapis.com/css?family=Raleway:400,300,600" rel="stylesheet" type="text/css">
    <link rel="stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/skeleton/2.0.4/skeleton.css"/>

    <!-- Necessary includes for LocusZoom.js -->
    <script src="dist/locuszoom.vendor.min.js" type="text/javascript"></script>
    <script src="dist/locuszoom.app.min.js" type="text/javascript"></script>
    <script type="application/javascript" src="dist/ext/lz-dynamic-urls.min.js"></script>

    <link rel="stylesheet" href="dist/locuszoom.css" type="text/css"/>

    <title>LocusZoom.js</title>

    <style>
      body {
        background-color: #FAFAFA;
        margin: 0px 20px;
      }
      img {
        max-width: 100%;
        box-sizing: border-box;
      }
      div.example > a > h6 {
        margin-bottom: 0.5em;
      }
      div.example > a > img {
        border: 1px solid #8A8A8A;
        margin-bottom: 1.4em;
      }
    </style>

  </head>

  <body>
    <div class="container">

      <h3 style="margin-top: 1em;"><strong>LocusZoom.js</strong></h3>
      <h6 style="color: #777">A LocusZoom extension track for interactively visualizing single-cell chromatin co-accessibility links & scores of sites in genome to predict cis-regulatory interactions</h6>

      <hr>

      <a name="try_it"></a>
      <div class="row">
        <div class="ten columns">
          <div id="ca-plot"></div>
        </div>
      </div>

      <div class="row">
        <footer style="text-align: center;">
          &copy; Copyright 2019 <a href="https://github.com/statgen">The University of Michigan Center for Statistical Genetics</a><br>
        </footer>
      </div>

    </div>

    <script type="text/javascript">

    // Determine if we're online, based on browser state or presence of an optional query parameter
    var online = !(typeof navigator != "undefined" && !navigator.onLine);
    if (window.location.search.indexOf("offline") != -1){ online = false; }

    // Define LocusZoom Data Sources object differently depending on online status
    var apiBase, data_sources;
    apiBase1 = "http://www.apps.t2depigenome.org/locuszoom/staticdata/";
    apiBase2 = "https://portaldev.sph.umich.edu/api/v1/";
    apiBase = window.location.origin + window.location.pathname.substr(0, window.location.pathname.lastIndexOf("/") + 1) + "staticdata/";
    data_sources = new LocusZoom.DataSources()
      .add("access", ["AccessibilityLZ", {url: apiBase + "co_accesibility.json", params: { build: 'GRCh37' }}])
      .add("gene", ["GeneLZ", { url: apiBase2 + "annotation/genes/", params: { build: 'GRCh37' } }]);

    // Get the standard association plot layout from LocusZoom's built-in layouts
    var stateUrlMapping = {chr: "chrom", start: "start", end: "end"};
    // Fetch initial position from the URL, or use some defaults
    var initialState = LocusZoom.ext.DynamicUrls.paramsFromUrl(stateUrlMapping);
    if (!Object.keys(initialState).length) {
        initialState = {chr: 11, start: 2182049, end: 2904920};
    }
    layout = LocusZoom.Layouts.get("plot", "coaccessibility", {state: initialState});
    layout.dashboard = LocusZoom.Layouts.get("dashboard", "region_nav_plot");


    // Generate the LocusZoom plot, and reflect the initial plot state in url
    window.plot = LocusZoom.populate("#ca-plot", data_sources, layout);
    // Add a "click gene to match accessibility loop" behavior
    plot.panels['genes'].data_layers["genes"].layout.match = { send: 'gene_name', receive: 'gene_name' };

    var color_config = [
        {
            field: 'lz_highlight_match', // Special field name whose presence triggers custom rendering
            scale_function: 'if',
            parameters: {
              field_value: true,
              then: '#ff0000',
            },
        },
        {
            field: 'lz_highlight_match', // Special field name whose presence triggers custom rendering
            scale_function: 'if',
            parameters: {
                field_value: false,
                then: '#EAE6E6',
            },
        },
        '#363696',
    ];
    plot.panels['genes'].data_layers["genes"].layout.color = color_config;
    plot.panels['genes'].data_layers["genes"].layout.stroke = color_config;

    // Changes in the plot can be reflected in the URL, and vice versa (eg browser back button can go back to
    //   a previously viewed region)
    LocusZoom.ext.DynamicUrls.plotUpdatesUrl(plot, stateUrlMapping);
    LocusZoom.ext.DynamicUrls.plotWatchesUrl(plot, stateUrlMapping);

    // Add a basic loader to each panel (one that shows when data is requested and hides when one rendering)
    plot.layout.panels.forEach(function(panel){
      plot.panels[panel.id].addBasicLoader();
    });

    // Create a method to parse a region string into a 600Kb genome range and load it
    function jumpTo(region) {
      var target = region.split(":");
      var chr = target[0];
      var pos = target[1];
      var start = 0;
      var end = 0;
      if (!pos.match(/[-+]/)) {
        start = +pos - 300000;
        end = +pos + 300000
      }
      plot.applyState({ chr: chr, start: start, end: end, ldrefvar: "" });
      return false;
    }


  </script>

  </body>
</html>
