<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link href="//fonts.googleapis.com/css?family=Raleway:400,300,600" rel="stylesheet" type="text/css">
    <link rel="stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/skeleton/2.0.4/skeleton.css"/>
    
    <!-- Necessary includes for LocusZoom.js -->
    <script src="../locuszoom.vendor.min.js" type="text/javascript"></script>
    <script src="../locuszoom.app.js" type="text/javascript"></script>
    <link rel="stylesheet" href="../locuszoom.css" type="text/css"/>
    
    <title>LocusZoom.js ~ PheWAS Demo</title>

    <style>
      body {
        background-color: #FAFAFA;
        margin: 0px 20px;
      }
      img {
        max-width: 100%;
        box-sizing: border-box;
      }
      pre {
        overflow: auto;
        background-color: #f1f1f1;
        border-radius: 3px;
        font-size: 1em;
      }
      code {
        font-size: 1em;
      }
      .spacer {
        padding: 0px 15px 0px 15px;
      }
      .required {
        color: red;
      }
    </style>

  </head>

  <body>
    <div class="container">

      <h1 style="margin-top: 1em;"><strong>LocusZoom.js</strong></h1>

      <h3 style="color: #777">PheWAS Demonstration</h3>

      <p>
        A PheWAS visualization displays association p-values between a given genetic variant and multiple phenotypes. 
        In this demonstration, we are showing single variant association statistics from multiple publicly available meta-analyses of genome-wide association studies (GWAS), stored locally in a database at the Center for Statistical Genetics (University of Michigan.)
        For the purposes of the demonstration, we chose reasonable groupings of traits into categories, but any arbitrary grouping is possible. 
      </p>

      <p>
        LocusZoom.js requires a REST API endpoint to retrieve data for the plot. The format of the response is documented <a target=_blank href='http://portaldev.sph.umich.edu/docs/api/v1/index.html#phewas-all-available-results-for-a-given-variant'>in our API documentation</a>, with a brief synopsis also included below.
      </p>

      <p>
        Example code for creating the plot can be found on the source for this page, and also exists within the <a target=_blank href='https://github.com/statgen/locuszoom/blob/master/examples/phewas_scatter.html'>locuszoom.js repository</a>.
      </p>

      <h3 style="color: #777">Plot</h3>

      <div class="row">
        <div class="two columns">
          <h5>Examples</h5>
          <ul class="top_hits" style="padding-left: 0.2rem; min-width: 110px;list-style-type: none">
            <li><a href="javascript:void(0);" onclick="loadPheWAS('10_114758349_C_T')"><strong>10:114758349_C/T (TCF7L2)</strong></a></li>
            <li><a href="javascript:void(0);" onclick="loadPheWAS('11_92690661_T_C')"><strong>11:92690661_T/C (MTNR1B)</strong></a></li>
            <li><a href="javascript:void(0);" onclick="loadPheWAS('2_227099180_C_T')"><strong>2:227099180_C/T (IRS1)</strong></a></li>
            <li><a href="javascript:void(0);" onclick="loadPheWAS('19_45422846_G_A')"><strong>19:45422846_G/A (APOE)</strong></a></li>
            <li><a href="javascript:void(0);" onclick="loadPheWAS('12_112591686_G_A')"><strong>12:112591686_G/A (TRAFD1)</strong></a></li>
            <li><a href="javascript:void(0);" onclick="loadPheWAS('11_61552680_G_T')"><strong>11:61552680_G/T (FADS3)</strong></a></li>
          </ul>
        </div>
        <div class="ten columns">
          <div id="plot"></div>
        </div>
      </div>

      <h3 style="color: #777">API</h3>

      <p>
        Full documentation/examples are available <a target=_blank href='http://portaldev.sph.umich.edu/docs/api/v1/index.html#phewas-all-available-results-for-a-given-variant'>on our website</a>. 
      </p>

      <p>
        LocusZoom.js will fire off an API request containing 3 parameters: <code>build</code>, <code>filter</code>, and <code>format</code>. 
      </p>

      <p><code>build</code> specifies the genome build for the given variant. An example would be 'GRCh37' or 'GRCh38'. The trailing version information (e.g. 'GRCh37p13.3') will not be present.</p>
      <p><code>filter</code> specifies the query which contains the variant of interest. An example filter string would be: <code>variant eq '11:61552680_G/T'</code>. </p>
      <p><code>format</code> specifies the format of the response. For PheWAS, it is always <code>objects</code>, which is an array of objects, each object containing the result for one association between variant and phenotype.</p>

      <p>Here is an example request that LocusZoom.js will generate:</p>

      <pre><div class="spacer">
http://portaldev.sph.umich.edu/api_internal_dev/v1/statistic/phewas/?build=GRCh37&format=objects&filter=variant eq '10:114758349_C/T'
      </div></pre>

      <p>The response must be in the following format:</p> 
      <pre><div class="spacer">
{
  <span class="required">"data"</span>: [
    {
      "analysis": "DIAGRAM 1000G T2D meta-analysis",
      "build": "GRCh37",
      "chromosome": "10",
      "id": 45,
      <span class="required">"log_pvalue"</span>: 107.032,
      "pmid": "28566273",
      "position": 114758349,
      "ref_allele": "C",
      "ref_allele_freq": null,
      "score_test_stat": null,
      "study": "DIAGRAM",
      "tech": null,
      "trait": "T2D",
      <span class="required">"trait_group"</span>: "Metabolic disease",
      <span class="required">"trait_label"</span>: "Type 2 diabetes",
      <span class="required">"variant"</span>: "10:114758349_C/T"
    }
  ],
  <span class="required">"lastPage"</span>: null,
  <span class="required">"meta"</span>: {
    <span class="required">"build"</span>: [
      "GRCh37"
    ]
  }
} 
      </div></pre>

      <p>
        Fields in <span class="required">red</span> above are required fields that must be present in the response.
        Extra fields can be included if desired (for example, <code>pmid</code> for the pubmed ID of the paper/analysis.)
        These extra fields can be used in the tooltip or other locations. 
      </p>

      <hr>

      <div class="row">
        <footer style="text-align: center;">
          &copy; Copyright 2017 <a href="https://github.com/statgen">The University of Michigan Center for Statistical Genetics</a><br>
        </footer>
      </div>

      <script type="text/javascript">

        // What variant are we using to display a PheWAS?

        var variantChrom = 10;
        var variantPosition = 114758349;
        var variantRefAlt = "C/T";
        // Variants are expressed in format `10:114758349_C/T`;
        var variant = [variantChrom, ":", variantPosition, "_", variantRefAlt].join("");

        // Genome base pairs static data
        var genome_data =  [
          { chr: 1, base_pairs: 249250621 },
          { chr: 2, base_pairs: 243199373 },
          { chr: 3, base_pairs: 198022430 },
          { chr: 4, base_pairs: 191154276 },
          { chr: 5, base_pairs: 180915260 },
          { chr: 6, base_pairs: 171115067 },
          { chr: 7, base_pairs: 159138663 },
          { chr: 8, base_pairs: 146364022 },
          { chr: 9, base_pairs: 141213431 },
          { chr: 10, base_pairs: 135534747 },
          { chr: 11, base_pairs: 135006516 },
          { chr: 12, base_pairs: 133851895 },
          { chr: 13, base_pairs: 115169878 },
          { chr: 14, base_pairs: 107349540 },
          { chr: 15, base_pairs: 102531392 },
          { chr: 16, base_pairs: 90354753 },
          { chr: 17, base_pairs: 81195210 },
          { chr: 18, base_pairs: 78077248 },
          { chr: 19, base_pairs: 59128983 },
          { chr: 20, base_pairs: 63025520 },
          { chr: 21, base_pairs: 48129895 },
          { chr: 22, base_pairs: 51304566 }
        ];

        // Define LocusZoom Data Sources object differently depending on online status
        var online = !(typeof navigator != "undefined" && !navigator.onLine);
        if (window.location.search.indexOf("offline") != -1){ online = false; }

        var apiBase;
        var dataSources= new LocusZoom.DataSources();
        if (online){
          apiBase = "https://portaldev.sph.umich.edu/api/v1/";
          dataSources
            .add("phewas", ["PheWASLZ", {
                // TODO: This source is currently development-only
                url: "https://portaldev.sph.umich.edu/" + "api_internal_dev/v1/statistic/phewas/",
                params: { build: ["GRCh37"] }
            }])
            .add("gene", ["GeneLZ", { url: apiBase + "annotation/genes/", params: {source: 2} }])
            .add("constraint", ["GeneConstraintLZ", { url: "http://exac.broadinstitute.org/api/constraint" }])

        } else {
          apiBase = window.location.origin + window.location.pathname.substr(0, window.location.pathname.lastIndexOf("/") + 1) + "../staticdata/";
          dataSources
            .add("phewas", ["PheWASLZ", {
                url: apiBase + "phewas_" + String(variantChrom) + "_" + String(variantPosition) + "_C-T.json",
                params: { build: ["GRCh37"] }
            }])
            .add("gene", ["GeneLZ", { url: apiBase + "genes-for-phewas.json", params: {source: 2} }])
            .add("constraint", ["GeneConstraintLZ", {  url: apiBase + "genes_" + String(variantChrom - 250000) + "-" + String(variantChrom + 250000) + ".json" }]);
        }
        // Static blobs (independent of host)
        dataSources
          .add("variant", ["StaticJSON", [{ "x": variantPosition, "y": 0 }, { "x": variantPosition, "y": 1 }]])
          .add("genome", ["StaticJSON", genome_data]);

        // Define the layout
        var mods = {
          state: {
            variant: variant,
            start: variantPosition - 250000,
            end: variantPosition + 250000,
            chr: 10
          }
        };
        var layout = LocusZoom.Layouts.get("plot", "standard_phewas", mods);
        layout.panels[0].margin.top = 32;
        layout.panels[0].data_layers[0].offset = 7.30103; // Higher offset for line of GWAS significance than the default 4.522
        layout.panels[2].data_layers.push({
          id: "variant",
          type: "orthogonal_line",
          orientation: "vertical",
          offset: variantPosition,
          style: {
            "stroke": "#FF3333",
            "stroke-width": "2px",
            "stroke-dasharray": "4px 4px"
          }
        });

        // Modify the tooltips for PheWAS result data layer points to contain more data. The fields in this sample
        //   tooltip are specific to the LZ-Portal API, and are not guaranteed to be in other PheWAS datasources.
        var phewas_layer = layout.panels[0].data_layers[1];
        // Tell the layer to also fetch some special fields; otherwise the datasource will hide this info (TODO)
        phewas_layer.fields.push("phewas:pmid", "phewas:analysis", "phewas:study");
        phewas_layer.tooltip.html = [
            "<strong>Trait:</strong> {{phewas:trait_label|htmlescape}}<br>",
            "<strong>Trait Category:</strong> {{phewas:trait_group|htmlescape}}<br>",
            "<strong>P-value:</strong> {{phewas:log_pvalue|logtoscinotation|htmlescape}}<br>",

            "{{#if phewas:study}}",
              "<br><strong>Study:</strong> {{phewas:study|htmlescape}}<br>",
            "{{/if}}",
            "{{#if phewas:pmid}} {{#if phewas:analysis}}",
                '<strong>Analysis:</strong> <a href="https://www.ncbi.nlm.nih.gov/pubmed/?term={{phewas:pmid}}">{{phewas:analysis|htmlescape}}</a>',
            "{{/if}} {{/if}}"
        ].join("");

        // Generate the plot
        var plot = LocusZoom.populate("#plot", dataSources, layout);
        plot.panels.phewas.setTitle("Variant " + variantChrom + ":" + String(variantPosition) + " " + variantRefAlt);

        // Function to load new PheWAS data into the plot
        function loadPheWAS(variant_tag){
          var variant = variant_tag.split("_");
          var vid = variant[0] + ":" + variant[1] + "_" + variant[2] + "/" + variant[3];
          var locus = +variant[1];
          var state = {
              variant: vid,
              start: (locus - 250000),
              end: (locus + 250000),
              chr: +variant[0]
          };
          plot.panels.genes.data_layers.variant.layout.offset = locus;
          plot.panels.phewas.setTitle("Variant " + variant[0] + ":" + variant[1] + " " + variant[2] + "/" + variant[3]);
          plot.applyState(state);
          plot.clearPanelData(null, "reset");
        }

      </script>

    </div>

  </body>
</html>
