<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link href="//fonts.googleapis.com/css?family=Raleway:400,300,600" rel="stylesheet" type="text/css">
    <link rel="stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/skeleton/2.0.4/skeleton.css"/>
    
    <!-- Necessary includes for LocusZoom.js -->
    <script src="../locuszoom.vendor.min.js" type="text/javascript"></script>
    <script src="../locuszoom.app.js" type="text/javascript"></script>
    <link rel="stylesheet" href="../locuszoom.css" type="text/css"/>
    
    <title>LocusZoom.js ~ PheWAS (Scatter) Example</title>

    <style>
      body {
        background-color: #FAFAFA;
        margin: 0px 20px;
      }
      img {
        max-width: 100%;
        box-sizing: border-box;
      }
    </style>

  </head>

  <body>
    <div class="container">

      <h1 style="margin-top: 1em;"><strong>Locuszoom.js</strong></h1>

      <h3 style="float: left; color: #777">PheWAS (Scatter) Example</h3>
      <h6 style="float: right;"><a href="../index.html">&lt; return home</a></h6>

      <hr style="clear: both;">

      <p>This example shows a basic approach to visualizing Phenome-Wide Association Study (PheWAS) results with a scatter plot. Data is sampled from the <a href="https://www.michigangenomics.org/genetic_data.html" target="_blank">Michigan Genomics Initiative</a> data set. Phenotypes are generated from EHR data by combining ICD-9 codes and are groups into broad categories to be listed as custom x-axis ticks. Otherwise points are plotted along the y-axis by their p-value, similar to how a "standard" LocusZoom plot would look.<p>
      <p>This plot also shows the LocusZoom.js label system, which renders label annotations on scatter points with logic to keep labels from running over each other. It also features a "genome legend" panel to provide genome-wide context for the variant cross-section represented by the data currently shown in the plot.</p>
      <p>Use the links to the left to see a selection of different loci with their corresponding PheWAS results.</p>

      <div class="row">
        <div class="two columns">
          <h4>Genes</h4>
          <ul class="top_hits" style="padding-left: 0.2rem; min-width: 110px;">
            <li><a href="javascript:void(0);" onclick="loadPheWAS('10_123352136_C_T')"><strong>FGFR2</strong><br><small>(Breast Cancer)</small></a></li>
            <li><a href="javascript:void(0);" onclick="loadPheWAS('4_111704536_A_T')"><strong>PITX2</strong><br><small>(Atrial Fibrillation)</small></a></li>
            <li><a href="javascript:void(0);" onclick="loadPheWAS('10_114758349_C_T')"><strong>TCF7L2</strong><br><small>(Type 2 Diabetes)</small></a></li>
            <li><a href="javascript:void(0);" onclick="loadPheWAS('1_169519049_T_C')"><strong>Factor 5</strong><br><small>(Thrombosis)</small></a></li>
            <li><a href="javascript:void(0);" onclick="loadPheWAS('19_45411941_T_C')"><strong>APOE</strong><br><small>(Dementia)</small></a></li>
          </ul>
        </div>
        <div class="ten columns">
          <div id="plot"></div>
        </div>
      </div>

      <hr>

      <div class="row">
        <footer style="text-align: center;">
          &copy; Copyright 2017 <a href="https://github.com/statgen">The University of Michigan Center for Statistical Genetics</a><br>
        </footer>
      </div>

      <script type="text/javascript">

        // What variant are we using to display a PheWAS?

        var variantChrom = 10;
        var variantPosition = 114758349;
        var variantRefAlt = "C/T";
        // Variants are expressed in format `10:114758349_C/T`;
        var variant = [variantChrom, ":", variantPosition, "_", variantRefAlt].join("");

        // Genome base pairs static data
        var genome_data =  [
          { chr: 1, base_pairs: 249250621 },
          { chr: 2, base_pairs: 243199373 },
          { chr: 3, base_pairs: 198022430 },
          { chr: 4, base_pairs: 191154276 },
          { chr: 5, base_pairs: 180915260 },
          { chr: 6, base_pairs: 171115067 },
          { chr: 7, base_pairs: 159138663 },
          { chr: 8, base_pairs: 146364022 },
          { chr: 9, base_pairs: 141213431 },
          { chr: 10, base_pairs: 135534747 },
          { chr: 11, base_pairs: 135006516 },
          { chr: 12, base_pairs: 133851895 },
          { chr: 13, base_pairs: 115169878 },
          { chr: 14, base_pairs: 107349540 },
          { chr: 15, base_pairs: 102531392 },
          { chr: 16, base_pairs: 90354753 },
          { chr: 17, base_pairs: 81195210 },
          { chr: 18, base_pairs: 78077248 },
          { chr: 19, base_pairs: 59128983 },
          { chr: 20, base_pairs: 63025520 },
          { chr: 21, base_pairs: 48129895 },
          { chr: 22, base_pairs: 51304566 }
        ];

        // Define LocusZoom Data Sources object differently depending on online status
        var online = !(typeof navigator != "undefined" && !navigator.onLine);
        if (window.location.search.indexOf("offline") != -1){ online = false; }

        var apiBase;
        var dataSources= new LocusZoom.DataSources();
        if (online){
          apiBase = "https://portaldev.sph.umich.edu/api/v1/";
          dataSources
            .add("phewas", ["PheWASLZ", {
                // TODO: This source is currently development-only
                url: "https://portaldev.sph.umich.edu/" + "api_internal_dev/v1/statistic/phewas/",
                params: { build: ["GRCh37"] }
            }])
            .add("gene", ["GeneLZ", { url: apiBase + "annotation/genes/", params: {source: 2} }])
            .add("constraint", ["GeneConstraintLZ", { url: "http://exac.broadinstitute.org/api/constraint" }])

        } else {
          apiBase = window.location.origin + window.location.pathname.substr(0, window.location.pathname.lastIndexOf("/") + 1) + "../staticdata/";
          dataSources
            .add("phewas", ["PheWASLZ", {
                url: apiBase + "phewas_" + String(variantChrom) + "_" + String(variantPosition) + "_C-T.json",
                params: { build: ["GRCh37"] }
            }])
            .add("gene", ["GeneLZ", { url: apiBase + "genes-for-phewas.json", params: {source: 2} }])
            .add("constraint", ["GeneConstraintLZ", {  url: apiBase + "genes_" + String(variantChrom - 250000) + "-" + String(variantChrom + 250000) + ".json" }]);
        }
        // Static blobs (independent of host)
        dataSources
          .add("variant", ["StaticJSON", [{ "x": variantPosition, "y": 0 }, { "x": variantPosition, "y": 1 }]])
          .add("genome", ["StaticJSON", genome_data]);

        // Define the layout
        var mods = {
          state: {
            variant: variant,
            start: variantPosition - 250000,
            end: variantPosition + 250000,
            chr: 10
          }
        };
        var layout = LocusZoom.Layouts.get("plot", "standard_phewas", mods);
        layout.panels[0].margin.top = 32;
        layout.panels[0].data_layers[0].offset = 7.30103; // Higher offset for line of GWAS significance than the default 4.522
        layout.panels[2].data_layers.push({
          id: "variant",
          type: "orthogonal_line",
          orientation: "vertical",
          offset: variantPosition,
          style: {
            "stroke": "#FF3333",
            "stroke-width": "2px",
            "stroke-dasharray": "4px 4px"
          }
        });

        // Modify the tooltips for PheWAS result data layer points to contain more data. The fields in this sample
        //   tooltip are specific to the LZ-Portal API, and are not guaranteed to be in other PheWAS datasources.
        var phewas_layer = layout.panels[0].data_layers[1];
        // Tell the layer to also fetch some special fields; otherwise the datasource will hide this info (TODO)
        phewas_layer.fields.push("phewas:pmid", "phewas:analysis", "phewas:study");
        phewas_layer.tooltip.html = [
            "<strong>Trait:</strong> {{phewas:trait_label|htmlescape}}<br>",
            "<strong>Trait Category:</strong> {{phewas:trait_group|htmlescape}}<br>",
            "<strong>P-value:</strong> {{phewas:log_pvalue|logtoscinotation|htmlescape}}<br>",

            "{{#if phewas:study}}",
              "<br><strong>Study:</strong> {{phewas:study|htmlescape}}<br>",
            "{{/if}}",
            "{{#if phewas:pmid}} {{#if phewas:analysis}}",
                '<strong>Analysis:</strong> <a href="https://www.ncbi.nlm.nih.gov/pubmed/?term={{phewas:pmid}}">{{phewas:analysis|htmlescape}}</a>',
            "{{/if}} {{/if}}"
        ].join("");

        // Generate the plot
        var plot = LocusZoom.populate("#plot", dataSources, layout);
        plot.panels.phewas.setTitle("Variant " + variantChrom + ":" + String(variantPosition) + " " + variantRefAlt);

        // Function to load new PheWAS data into the plot
        function loadPheWAS(variant_tag){
          var variant = variant_tag.split("_");
          var vid = variant[0] + ":" + variant[1] + "_" + variant[2] + "/" + variant[3];
          var locus = +variant[1];
          var state = {
              variant: vid,
              start: (locus - 250000),
              end: (locus + 250000),
              chr: +variant[0]
          };
          plot.panels.genes.data_layers.variant.layout.offset = locus;
          plot.panels.phewas.setTitle("Variant " + variant[0] + ":" + variant[1] + " " + variant[2] + "/" + variant[3]);
          plot.applyState(state);
          plot.clearPanelData(null, "reset");
        }

      </script>

    </div>

  </body>
</html>
